sdks/java/harness/src/main/java/org/apache/beam/fn/harness/WindowMappingFnRunner.java

```diff
@@ -0,0 +1,125 @@
+  static final String URN = validateCommonUrn("beam:transform:map_windows:v1");
+
+  /**
+   * A registrar which provides a factory to handle mapping main input windows onto side input
+   * windows.
+   */
+  @AutoService(PTransformRunnerFactory.Registrar.class)
+  public static class Registrar implements PTransformRunnerFactory.Registrar {
+
+    @Override
+    public Map<String, PTransformRunnerFactory> getPTransformRunnerFactories() {
+      return ImmutableMap.of(URN, new Factory());
+    }
+  }
+
+  /** A factory for {@link BeamFnDataReadRunner}s. */
+  static class Factory<InputT extends BoundedWindow, OutputT extends BoundedWindow>
+      implements PTransformRunnerFactory<WindowMappingFnRunner<InputT, OutputT>> {
+
+    @Override
+    public WindowMappingFnRunner<InputT, OutputT> createRunnerForPTransform(
+        PipelineOptions pipelineOptions,
+        BeamFnDataClient beamFnDataClient,
+        BeamFnStateClient beamFnStateClient,
+        String pTransformId,
+        PTransform pTransform,
+        Supplier<String> processBundleInstructionId,
+        Map<String, PCollection> pCollections,
+        Map<String, RunnerApi.Coder> coders,
+        Map<String, RunnerApi.WindowingStrategy> windowingStrategies,
+        Multimap<String, FnDataReceiver<WindowedValue<?>>> pCollectionIdsToConsumers,
+        Consumer<ThrowingRunnable> addStartFunction,
+        Consumer<ThrowingRunnable> addFinishFunction) throws IOException {
+
+      Collection<FnDataReceiver<WindowedValue<OutputT>>> consumers =
+          (Collection) pCollectionIdsToConsumers.get(
+              getOnlyElement(pTransform.getOutputsMap().values()));
+
+      SdkFunctionSpec windowMappingFnPayload =
+          SdkFunctionSpec.parseFrom(pTransform.getSpec().getPayload());
+      WindowMappingFn<OutputT> windowMappingFn =
+          (WindowMappingFn<OutputT>) PCollectionViewTranslation.windowMappingFnFromProto(
+              windowMappingFnPayload);
+
+      WindowMappingFnRunner<InputT, OutputT> runner =
+          new WindowMappingFnRunner<>(
+              windowMappingFn,
+              MultiplexingFnDataReceiver.forConsumers(consumers));
+      pCollectionIdsToConsumers.put(
+          Iterables.getOnlyElement(pTransform.getInputsMap().values()),
+          (FnDataReceiver) (FnDataReceiver<WindowedValue<InputT>>) runner::mapWindow);
+      return runner;
+    }
+  }
+
+  private final WindowMappingFn<OutputT> windowMappingFn;
+  private final FnDataReceiver<WindowedValue<OutputT>> consumer;
+
+  public WindowMappingFnRunner(
+      WindowMappingFn<OutputT> windowMappingFn,
+      FnDataReceiver<WindowedValue<OutputT>> consumer) {
+    this.windowMappingFn = windowMappingFn;
+    this.consumer = consumer;
+  }
+
+  public void mapWindow(WindowedValue<InputT> element) throws Exception {
+    WindowedValue<OutputT> output =
+        element.withValue(windowMappingFn.getSideInputWindow(element.getValue()));
```

Comment:
We do "withValue" just because we must accept/return windowed values, right? The fact that InputT and OutputT are windows themselves, seems a bit obscured here.

Reply:
Yeah, its an unfortunate thing that we currently require a WindowedValue everywhere when instead we could pass around an element context which is windowless. This was originally the goal of WindowedValue but because of its proliferation was abused.